File: medium/String/palindromic-substrings/palindromic_substrings.cpp
#include <string>
using namespace std;

class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();
        int count = 0;

        auto expand = [&](int l, int r) {
            while (l >= 0 && r < n && s[l] == s[r]) {
                count++;
                l--;
                r++;
            }
        };

        for (int i = 0; i < n; ++i) {
            expand(i, i);
            expand(i, i+1);
        }
        return count;
    }
};

// Test
#include <iostream>
int main() {
    Solution sol;
    cout << sol.countSubstrings("abc") << endl;
    cout << sol.countSubstrings("aaa") << endl;
    return 0;
}

File: medium/String/palindromic-substrings/palindromic_substrings.py
def countSubstrings(s: str) -> int:
    """
    Count the number of palindromic substrings in a string.
    Approach: Expand Around Center
    Time: O(n²)
    Space: O(1)
    """
    n = len(s)
    count = 0

    def expand(l, r):
        nonlocal count
        while l >= 0 and r < n and s[l] == s[r]:
            count += 1
            l -= 1
            r += 1

    for i in range(n):
        expand(i, i)    # odd
        expand(i, i+1) # even

    return count

if __name__ == "__main__":
    print(countSubstrings("abc"))  # 3
    print(countSubstrings("aaa"))  # 6
